### PR TITLE
[TG-2147] Documentation/review cbmc docs

### PR DESCRIPTION
**Pull request by Charliemowood**
The overall aim of this PR is to review the current documentation and improve it so that it is in a position that it can be published regularly via CI. 

Please look at the separate commits:

The focus of the first commit is to improve the layout of the table of contents.
The second commit is to make the front page more informative.
The third is separating the SATABS content from the CBMC user manual.

### CODE REVIEW

**Comment by Charliemowood** [comment]
The layout is look like this: 
screen shot 2018-01-30 at 09 51 01

**Comment by tautschnig** [comment]
Could you please rebase to avoid merge commits in the pull request? Thanks!

File: doc/architectural/cbmc-guide.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 \ingroup module_hidden
-\page cbmc-guide CProver Developer Manual
+\page cbmc-guide CBMC Developer Manual
```
Comment (tautschnig) [inline doc/architectural/cbmc-guide.md]: Nit pick: should live in the previous commit

File: doc/architectural/front-page.md [hunk 2]
```diff
@@ -5,15 +5,59 @@ CProver Developer Documentation
...
+CProver codebase. CProver is a platform for software verification.  Users can
+download CProver tools from the <a href="http://www.cprover.org/">CProver
+website</a>; contributors should use the
+<a href="https://github.com/diffblue/cbmc">repository</a> hosted on GitHub. CBMC
+is part of CProver.
+
+CBMC is a Bounded Model Checker for C and C++ programs. It supports C89, C99,
+most of C11 and most compiler extensions provided by gcc and Visual Studio. It
+also supports SystemC using Scoot. It allows verifying array bounds (buffer
+overflows), pointer safety, exceptions and user-specified assertions.
```
Comment (tautschnig) [inline doc/architectural/front-page.md]: I'd say "arithmetic exceptions"

File: doc/architectural/howto.md [hunk 3]
```diff
@@ -1,5 +1,5 @@
 \ingroup module_hidden
-\page tutorial CProver Developer Tutorial
+\page tutorial CBMC Developer Tutorial
```
Comment (tautschnig) [inline doc/architectural/howto.md]: As above: should go with the previous commit.

**Review by tautschnig — COMMENTED** [review]
Looks good to me, except for the rebase to be done.

**Comment by Charliemowood** [comment]
@tautschnig thanks for your thoughts. Now rebased and your docs change suggestion implemented.

**Review by tautschnig — APPROVED** [review]

File: doc/SATABS.md [hunk 4]
```diff
@@ -0,0 +1,696 @@
...
+property. If such a counterexample is found, it is presented to the
+engineer to facilitate localization and repair of the program.
+
+\subsection man_install-satabs Installing SATABS
+
+### Requirements
+
+SATABS is available for Windows, i86 Linux, and MacOS X. SATABS requires
+a code pre-processing environment comprising of a suitable preprocessor
+and an a set of header files.
```
Comment (romainbrenguier) [inline doc/SATABS.md]: "and a set"

File: doc/SATABS.md [hunk 5]
```diff
@@ -0,0 +1,696 @@
...
+Just as CBMC, SATABS attempts to build counterexamples that refute the
+property. If such a counterexample is found, it is presented to the
+engineer to facilitate localization and repair of the program.
+
+\subsection man_install-satabs Installing SATABS
+
+### Requirements
+
+SATABS is available for Windows, i86 Linux, and MacOS X. SATABS requires
+a code pre-processing environment comprising of a suitable preprocessor
```
Comment (romainbrenguier) [inline doc/SATABS.md]: "comprising a" or "including a"

File: doc/SATABS.md [hunk 6]
```diff
@@ -0,0 +1,696 @@
...
+    environment variable. Use the option
+
+        --modelchecker nusmv
+
+    to instruct SATABS to use NuSMV.
+
+-   **BOPPO**. Available from http://www.cprover.org/boppo/. BOPPO is
+    a model checker that uses SAT-solving algorithms. BOPPO relies on a
+    built-in SAT solver and Quantor, a solver for quantified boolean
+    formulas that is currently bundled with BOPPO, but also available
```
Comment (romainbrenguier) [inline doc/SATABS.md]: nitpick: I would use which rather than that (in both occurrences)

File: doc/SATABS.md [hunk 7]
```diff
@@ -0,0 +1,696 @@
...
+### Requirements
+
+We provide a graphical user interface to CBMC and SATABS, which is
+realized as a plugin to the Eclipse framework. Eclipse is available at
+http://www.eclipse.org. We do not provide installation instructions for
+Eclipse (basically, you only have to download the current version and
+extract the files to your hard-disk) and assume that you have already
+installed the current version.
+
+CBMC and SATABS have their own requirements. As an example, both CBMC and SATABS require a suitable preprocessor and a set of header files. As
```
Comment (romainbrenguier) [inline doc/SATABS.md]: long line

File: doc/SATABS.md [hunk 8]
```diff
@@ -0,0 +1,696 @@
...
+      dummy_release (&inode, &my_file);
+      lock_held = FALSE;
+      break;
+    default:
+      break;
+    }
+```
+
+The variable `random` is assigned non-deterministically. Subsequently,
+the value of `random` is restricted to be 0 &le `random` ≤ 3 by a call
```
Comment (romainbrenguier) [inline doc/SATABS.md]: strange to have once &le and once ≤ do they both work well with doxygen?

File: doc/SATABS.md [hunk 9]
```diff
@@ -0,0 +1,696 @@
...
+that the properties for this program hold on your computer with a
+specific set of en­vi­ron­ment variables. We have to assume that
+`environ` contains en­vi­ron­ment variables that have an arbitrary
+content of arbitrary length. The content of en­vi­ron­ment variables is
+not only arbitrary but could be malefic, since it can be modified by the
+user. The approximation of the behavior of `getenv` that is shipped with
+SATABS completely ignores the content of the string.
+
+Now let us have another look at the properties that SATABS generates for
+the models of the the string library and for `getenv`. Most of these
```
Comment (romainbrenguier) [inline doc/SATABS.md]: "of the string library"

**Review by romainbrenguier — APPROVED** [review]

File: doc/SATABS.md [hunk 10]
```diff
@@ -0,0 +1,696 @@
...
+
+1.  **Linux:** the preprocessor and the header files typically come with
+    a package called *gcc*, which must be installed prior to the
+    installation of SATABS.
+2.  **Windows:** The Windows version of SATABS requires the preprocessor
+    `cl.exe`, which is part of Visual Studio (including the free [Visual
+    Studio
+    Express](http://msdn.microsoft.com/vstudio/express/visualc/)).
+3.  **MacOS:** Install
+    [XCode](http://developer.apple.com/technologies/xcode.html) prior to
```
Comment (chrisr-diffblue) [inline doc/SATABS.md]: Do we actually need full XCode installed for any of CBMC? Or do we just need the "XCode command line tools", which is a significantly more lightweight package to install? I think just the command line tools can be installed with 'xcode-select --install' ?

File: doc/architectural/cbmc-guide.md [hunk 11]
```diff
@@ -1,5 +1,5 @@
 \ingroup module_hidden
-\page cbmc-guide CBMC Guide
+\page cbmc-guide CBMC Developer Manual
```
Comment (chrisr-diffblue) [inline doc/architectural/cbmc-guide.md]: If the document title has changed (which I approve of) is it worth changing the file name as well to match?

**Review by chrisr-diffblue — APPROVED** [review]
Looks good - thanks for tidying this up! A couple of non-blocking comments, but won't hold this back just for those.

**Comment by Charliemowood** [comment]
@romainbrenguier and @chrisr-diffblue thanks for the feedback. Implemented those ideas.

**Review by peterschrammel — CHANGES_REQUESTED** [review]
@Charliemowood, please implement the structural changes and clean-ups as discussed in person.

File: doc/architectural/front-page.md [hunk 12]
```diff
@@ -43,13 +41,16 @@ Contributing to the code base
 2. Clone the repository `git clone git@github.com:YOURNAME/cbmc.git`
 3. Create a branch from the `develop` branch (default branch)
```
Comment (chrisr-diffblue) [inline doc/architectural/front-page.md]: Points 1 and 2 should probably both be updated a touch too... I'd suggest something like:
1. For the CBMC repository on github
2. Clone your fork with `git clone......`

File: doc/architectural/front-page.md [hunk 13]
```diff
@@ -63,23 +64,23 @@ Overview of Documentation
 
 ### For contributors:
 
+* The \subpage cbmc-guide "CBMC Developer guide" is a single document describing
+  the layout of the codebase and many of the important data structures.
+  It probably contains more information than the module pages at the
+  moment, but may be somewhat out-of-date.
+
+* For higher-level architectural information, each of the pages under
+  the <a href="modules.html">Modules</a>.
```
Comment (chrisr-diffblue) [inline doc/architectural/front-page.md]: Erroneous full stop at the end of this line - should be a comma?

**Comment by Charliemowood** [comment]
@chrisr-diffblue thanks for the additional comments. Updated.

**Comment by Charliemowood** [comment]
@peterschrammel the _hidden has been brought back in to deal with toc problems. This is by no means perfect but the overall structural changes we discuss are now in this. Could you please review again?

**Review by peterschrammel — APPROVED** [review]

**Comment by Charliemowood** [comment]
👍

**Pull request merged**